trillian/ctdns/ctdns_server/main.go

```diff
@@ -0,0 +1,250 @@
+	if len(*rpcBackend) > 0 {
+		cfg, err = readCfg(*logConfig, *rpcBackend)
+	} else {
+		cfg, err = readMultiCfg(*logConfig)
+	}
+
+	if err != nil {
+		glog.Exitf("Failed to read config: %v", err)
+	}
+
+	beMap, err := ctfe.ValidateLogMultiConfig(cfg)
+	if err != nil {
+		glog.Exitf("Invalid config: %v", err)
+	}
+
+	glog.CopyStandardLogTo("WARNING")
+	glog.Info("**** CT DNS Server Starting ****")
+
+	var res naming.Resolver
+	if len(*etcdServers) > 0 {
+		// Use etcd to provide endpoint resolution.
+		cfg := clientv3.Config{Endpoints: strings.Split(*etcdServers, ","), DialTimeout: 5 * time.Second}
+		client, err := clientv3.New(cfg)
+		if err != nil {
+			glog.Exitf("Failed to connect to etcd at %v: %v", *etcdServers, err)
+		}
+		etcdRes := &etcdnaming.GRPCResolver{Client: client}
+		res = etcdRes
+
+		// Also announce ourselves.
+		updateMetrics := naming.Update{Op: naming.Add, Addr: *metricsEndpoint}
+		glog.Infof("Announcing our presence in %v with %+v", *etcdMetricsService, updateMetrics)
+		etcdRes.Update(ctx, *etcdMetricsService, updateMetrics)
+
+		byeMetrics := naming.Update{Op: naming.Delete, Addr: *metricsEndpoint}
+		defer func() {
+			glog.Infof("Removing our presence in %v with %+v", *etcdMetricsService, byeMetrics)
+			etcdRes.Update(ctx, *etcdMetricsService, byeMetrics)
+		}()
+	} else {
+		// Use a fixed endpoint resolution that just returns the addresses configured on the command line.
+		res = util.FixedBackendResolver{}
+	}
+
+	// Dial all our log backends.
+	clientMap := make(map[string]trillian.TrillianLogClient)
+	for _, be := range beMap {
+		glog.Infof("Dialling backend: %v", be)
+		bal := grpc.RoundRobin(res)
+		opts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithBalancer(bal)}
+		if len(beMap) == 1 {
+			// If there's only one of them we use the blocking option as we can't
+			// serve anything until connected.
+			opts = append(opts, grpc.WithBlock())
+		}
+		conn, err := grpc.Dial(be.BackendSpec, opts...)
+		if err != nil {
+			glog.Exitf("Could not dial RPC udpServer: %v: %v", be, err)
+		}
+		defer conn.Close()
+		clientMap[be.Name] = trillian.NewTrillianLogClient(conn)
+	}
+
+	// Register DNS handlers for all the configured logs using the correct RPC
+	// client. Ignore any that don't specify a zone.
+	var zones int
+	for _, c := range cfg.LogConfigs.Config {
+		if len(c.DnsZone) > 0 {
+			zones++
+			if err := setupDNSHandler(clientMap[c.LogBackendName], *rpcDeadline, c); err != nil {
+				glog.Exitf("Failed to set up DNS log instance for %+v: %v", cfg, err)
+			}
+		}
+	}
+
+	if zones == 0 {
+		glog.Fatalf("No logs have a dns_zone configured. Exiting.")
+	}
+
+	// Handle metrics on the DefaultServeMux. We don't serve HTTP requests
+	// for clients, only DNS.
+	http.Handle("/metrics", promhttp.Handler())
+
+	// Return a 200 on the root, for GCE default health checking :/
+	http.HandleFunc("/", func(resp http.ResponseWriter, req *http.Request) { resp.WriteHeader(http.StatusOK) })
+
+	if *getSTHInterval > 0 {
```

Comment:
Could just drop this, given there currently aren't any metrics to get updated?  (I guess it depends on if/when you expect to come back and add metrics)

Reply:
I think it's needed for the get STH logic, which uses the same code as CTFE. I also plan to add metrics fairly soon.